zigzag-circuit.md

```diff
@@ -0,0 +1,122 @@
+
+- `replicaId` is a unique replica identifier (see the Filecoin Proofs spec for details)
+- `CommD` is the Merkle Tree root hash of the input data to the first layer
+- `CommR_{i}` is the Merkle Tree hash of the output of the DRG encoding at each layer `i` 
+- `CommRStar` is the hash of the concatenation of the `replicaId` and all the `commR`s.
+
+In the PoRep offline proof as presented in Fisch2019, the proof is large for blockchain usage. We use SNARKs to generate a proof of knowledge of an offline proof. In other words, we use encode the ZigZag proof verification algorithm in an arithmetic circuit and use SNARKs to prove that it was evaluated correctly.
+
+This circuit proves that given a Merkle root `CommD`, `CommR_l`, and `commRStar`, that the prover knew the correct replicated data at each layer.
+
+## Offline PoRep circuit
+
+**Public Parameters**: *Parameters that are embeded in the circuits or used to generate the circuit*
+
+- `LAYERS : UInt`: Number of layers
+- `LAYER_CHALLENGES[0..LAYERS]`: Number of challenges per layer
+- `EXPANSION_DEGREE: UInt`: Degree of the bipartite expander graph to extend dependencies between layers
+- `TREE_DEPTH: UInt`: Height of the Merkle tree
+
+**Public Inputs**: *Inputs that the prover uses to generate a SNARK proof and that the verifier uses to verify it*
+
+- `ReplicaId : Fr`: A unique identifier for the replica Id
+- `CommD : Fr`: the Merkle tree root hash of the original data (input to the first layer)
+- `CommRLast : Fr`: the Merkle tree root hash of the output of the last layer 
+- `CommRStar : Fr`: the aggregate of each layer's Merkle tree root hash
+- Inclusion paths: Binary representation of the Merkle tree path that must be proven packed into a single `Fr` element. We have the following inclusion paths:
+  - `InclusionPaths_{i=0..LAYERS}_{0..LAYER_CHALLENGES[i]}`: At each layer `i` we have `LAYER_CHALLENGES[i]` inclusion paths.
+  - `ParentsInclusionPaths_{l=0..LAYERS}_{c=0..LAYER_CHALLENGES[i]}_{0..EXPANSION_DEGREE+ BASE_DEGREE}`: At each layer `l` we have `LAYER_CHALLENGES[i]` an inclusion path for each parent node of the corresponding `InclusionPaths_{l}_{c}`.
+
+**Private Inputs**: *Inputs that the prover uses to generate a SNARK proof, these are not needed by the verifier to verify the proof*
+
+- `CommR_{i=0..LAYERS-1}`: Commitment of the the encoded data at layer `i`. 
+- Inclusion Proof: For each inclusion path in the public inputs, we provide a Merkle Tree path
+  - `InclusionHash_{i=0..LAYERS}_{0..LAYER_CHALLENGES[i]}_{0..TREE_DEPTH-1}`
+  - `ReplicaInclusionHash_{i=0..LAYERS}_{0..LAYER_CHALLENGES[i]}_{0..TREE_DEPTH-1}`
+  - `ParentInclusionHash_{l=0..LAYERS}_{c=0..LAYER_CHALLENGES[i]}_{0..EXPANSION_DEGREE+ BASE_DEGREE}_{0..TREE_DEPTH-1}`
+  - `InclusionLeaf_{i=0..LAYERS}_{0..LAYER_CHALLENGES[i]}`
```

Comment:
could we add some text to the private inputs?

Reply:
fixing